File: family-history-intake-form/node/src/intakeSubmission.ts
/**
 * Domain value object + parser for the intake form's submission.
 *
 * Pure: takes a raw form body (as Express populates it via urlencoded)
 * and returns either a validated IntakeSubmission or throws an
 * IntakeValidationError naming the field at fault.
 */

export type BiologicalSex = 'female' | 'male' | 'unknown';

export type SiblingRelation = 'sister' | 'brother' | 'half_sister' | 'half_brother';

export interface RelativeEntry {
  readonly displayName: string;
  readonly yearOfBirth?: number;
}

export interface SiblingEntry extends RelativeEntry {
  readonly relation: SiblingRelation;
  readonly biologicalSex: BiologicalSex;
}

export interface IntakeSubmission {
  readonly proband: {
    readonly displayName: string;
    readonly biologicalSex: BiologicalSex;
    readonly yearOfBirth?: number;
  };
  readonly mother?: RelativeEntry;
  readonly father?: RelativeEntry;
  readonly maternalGrandmother?: RelativeEntry;
  readonly maternalGrandfather?: RelativeEntry;
  readonly paternalGrandmother?: RelativeEntry;
  readonly paternalGrandfather?: RelativeEntry;
  readonly siblings: readonly SiblingEntry[];
}

export class IntakeValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'IntakeValidationError';
  }
}

const MIN_YEAR = 1850;
const MAX_YEAR = 2030;
const MAX_SIBLINGS = 4;
const KNOWN_SEXES: readonly BiologicalSex[] = ['female', 'male', 'unknown'];
const KNOWN_SIBLING_RELATIONS: readonly SiblingRelation[] = [
  'sister',
  'brother',
  'half_sister',
  'half_brother',
];

export function parseIntakeSubmission(body: Record<string, unknown>): IntakeSubmission {
  const probandName = requireNonEmpty(body, 'proband_name', "The patient's name");
  const probandSex = parseSex(optionalString(body, 'proband_sex'));
  const probandYear = parseYear(optionalString(body, 'proband_year'), 'proband_year');

  return {
    proband: {
      displayName: probandName,
      biologicalSex: probandSex,
      ...(probandYear !== undefined ? { yearOfBirth: probandYear } : {}),
    },
    ...optionalRelative(body, 'mother', 'mother'),
    ...optionalRelative(body, 'father', 'father'),
    ...optionalRelative(body, 'maternal_grandmother', 'maternalGrandmother'),
    ...optionalRelative(body, 'maternal_grandfather', 'maternalGrandfather'),
    ...optionalRelative(body, 'paternal_grandmother', 'paternalGrandmother'),
    ...optionalRelative(body, 'paternal_grandfather', 'paternalGrandfather'),
    siblings: parseSiblings(body),
  };
}

function optionalRelative(
  body: Record<string, unknown>,
  formPrefix: string,
  fieldName: string,
): Record<string, RelativeEntry> {
  const name = optionalString(body, `${formPrefix}_name`).trim();
  if (name === '') {
    return {};
  }
  const year = parseYear(optionalString(body, `${formPrefix}_year`), `${formPrefix}_year`);
  const entry: RelativeEntry = year === undefined ? { displayName: name } : { displayName: name, yearOfBirth: year };
  return { [fieldName]: entry };
}

function parseSiblings(body: Record<string, unknown>): readonly SiblingEntry[] {
  const siblings: SiblingEntry[] = [];
  for (let index = 0; index < MAX_SIBLINGS; index += 1) {
    const name = optionalString(body, `sibling_${index.toString()}_name`).trim();
    if (name === '') {
      continue;
    }
    const relation = parseSiblingRelation(
      optionalString(body, `sibling_${index.toString()}_relation`),
      index,
    );
    const year = parseYear(
      optionalString(body, `sibling_${index.toString()}_year`),
      `sibling_${index.toString()}_year`,
    );
    siblings.push({
      displayName: name,
      relation,
      biologicalSex: sexForSiblingRelation(relation),
      ...(year !== undefined ? { yearOfBirth: year } : {}),
    });
  }
  return siblings;
}

function parseSex(raw: string): BiologicalSex {
  if (raw === '') {
    return 'unknown';
  }
  const match = KNOWN_SEXES.find(known => known === raw);
  if (match === undefined) {
    throw new IntakeValidationError(`Unknown biological sex: ${raw}`);
  }
  return match;
}

function parseSiblingRelation(raw: string, index: number): SiblingRelation {
  const match = KNOWN_SIBLING_RELATIONS.find(known => known === raw);
  if (match === undefined) {
    throw new IntakeValidationError(
      `Sibling ${(index + 1).toString()} must have a relation (sister / brother / half_sister / half_brother).`,
    );
  }
  return match;
}

function sexForSiblingRelation(relation: SiblingRelation): BiologicalSex {
  return relation === 'sister' || relation === 'half_sister' ? 'female' : 'male';
}

function parseYear(raw: string, fieldName: string): number | undefined {
  const trimmed = raw.trim();
  if (trimmed === '') {
    return undefined;
  }
  const parsed = Number.parseInt(trimmed, 10);
  if (!Number.isFinite(parsed) || parsed < MIN_YEAR || parsed > MAX_YEAR) {
    throw new IntakeValidationError(
      `Field '${fieldName}' must be a year between ${MIN_YEAR.toString()} and ${MAX_YEAR.toString()}.`,
    );
  }
  return parsed;
}

function requireNonEmpty(body: Record<string, unknown>, key: string, label: string): string {
  const value = optionalString(body, key).trim();
  if (value === '') {
    throw new IntakeValidationError(`${label} is required.`);
  }
  return value;
}

function optionalString(body: Record<string, unknown>, key: string): string {
  const value = body[key];
  return typeof value === 'string' ? value : '';
}

File: family-history-intake-form/dotnet/src/FamilyIntake/IntakeSubmission.cs
namespace FamilyIntake;

public enum BiologicalSex
{
    Female,
    Male,
    Unknown,
}

public enum SiblingRelation
{
    Sister,
    Brother,
    HalfSister,
    HalfBrother,
}

public sealed record RelativeEntry(string DisplayName, int? YearOfBirth = null);

public sealed record SiblingEntry(
    string DisplayName,
    SiblingRelation Relation,
    BiologicalSex BiologicalSex,
    int? YearOfBirth = null);

public sealed record ProbandEntry(
    string DisplayName,
    BiologicalSex BiologicalSex,
    int? YearOfBirth = null);

public sealed record IntakeSubmission(
    ProbandEntry Proband,
    IReadOnlyList<SiblingEntry> Siblings,
    RelativeEntry? Mother = null,
    RelativeEntry? Father = null,
    RelativeEntry? MaternalGrandmother = null,
    RelativeEntry? MaternalGrandfather = null,
    RelativeEntry? PaternalGrandmother = null,
    RelativeEntry? PaternalGrandfather = null);

public sealed class IntakeValidationException : Exception
{
    public IntakeValidationException(string message) : base(message) { }
}

public static class IntakeSubmissionParser
{
    private const int MinYear = 1850;
    private const int MaxYear = 2030;
    private const int MaxSiblings = 4;

    public static IntakeSubmission Parse(IReadOnlyDictionary<string, string> body)
    {
        var probandName = RequireNonEmpty(body, "proband_name", "The patient's name");
        var probandSex = ParseSex(OptionalString(body, "proband_sex"));
        var probandYear = ParseYear(OptionalString(body, "proband_year"), "proband_year");

        return new IntakeSubmission(
            Proband: new ProbandEntry(probandName, probandSex, probandYear),
            Siblings: ParseSiblings(body),
            Mother: OptionalRelative(body, "mother"),
            Father: OptionalRelative(body, "father"),
            MaternalGrandmother: OptionalRelative(body, "maternal_grandmother"),
            MaternalGrandfather: OptionalRelative(body, "maternal_grandfather"),
            PaternalGrandmother: OptionalRelative(body, "paternal_grandmother"),
            PaternalGrandfather: OptionalRelative(body, "paternal_grandfather"));
    }

    private static RelativeEntry? OptionalRelative(IReadOnlyDictionary<string, string> body, string formPrefix)
    {
        var name = OptionalString(body, $"{formPrefix}_name").Trim();
        if (name.Length == 0)
        {
            return null;
        }
        var year = ParseYear(OptionalString(body, $"{formPrefix}_year"), $"{formPrefix}_year");
        return new RelativeEntry(name, year);
    }

    private static List<SiblingEntry> ParseSiblings(IReadOnlyDictionary<string, string> body)
    {
        var siblings = new List<SiblingEntry>();
        for (var index = 0; index < MaxSiblings; index += 1)
        {
            var name = OptionalString(body, $"sibling_{index}_name").Trim();
            if (name.Length == 0)
            {
                continue;
            }
            var relation = ParseSiblingRelation(OptionalString(body, $"sibling_{index}_relation"), index);
            var year = ParseYear(OptionalString(body, $"sibling_{index}_year"), $"sibling_{index}_year");
            siblings.Add(new SiblingEntry(name, relation, SexForSiblingRelation(relation), year));
        }
        return siblings;
    }

    private static BiologicalSex ParseSex(string raw) => raw switch
    {
        "" => BiologicalSex.Unknown,
        "female" => BiologicalSex.Female,
        "male" => BiologicalSex.Male,
        "unknown" => BiologicalSex.Unknown,
        _ => throw new IntakeValidationException($"Unknown biological sex: {raw}"),
    };

    private static SiblingRelation ParseSiblingRelation(string raw, int index) => raw switch
    {
        "sister" => SiblingRelation.Sister,
        "brother" => SiblingRelation.Brother,
        "half_sister" => SiblingRelation.HalfSister,
        "half_brother" => SiblingRelation.HalfBrother,
        _ => throw new IntakeValidationException(
            $"Sibling {index + 1} must have a relation (sister / brother / half_sister / half_brother)."),
    };

    private static BiologicalSex SexForSiblingRelation(SiblingRelation relation) => relation switch
    {
        SiblingRelation.Sister or SiblingRelation.HalfSister => BiologicalSex.Female,
        SiblingRelation.Brother or SiblingRelation.HalfBrother => BiologicalSex.Male,
        _ => throw new ArgumentOutOfRangeException(nameof(relation)),
    };

    private static int? ParseYear(string raw, string fieldName)
    {
        var trimmed = raw.Trim();
        if (trimmed.Length == 0)
        {
            return null;
        }
        if (!int.TryParse(trimmed, out var parsed) || parsed < MinYear || parsed > MaxYear)
        {
            throw new IntakeValidationException(
                $"Field '{fieldName}' must be a year between {MinYear} and {MaxYear}.");
        }
        return parsed;
    }

    private static string RequireNonEmpty(IReadOnlyDictionary<string, string> body, string key, string label)
    {
        var value = OptionalString(body, key).Trim();
        if (value.Length == 0)
        {
            throw new IntakeValidationException($"{label} is required.");
        }
        return value;
    }

    private static string OptionalString(IReadOnlyDictionary<string, string> body, string key)
    {
        return body.TryGetValue(key, out var value) ? value : string.Empty;
    }
}
